internal/provider/data_source.go

```diff
@@ -160,66 +184,67 @@ The program must also be executable according to the platform where Terraform is
-
-			return diag.Diagnostics{
-				{
-					Severity: diag.Error,
-					Summary:  "External Program Execution Failed",
-					Detail: "The data source received an unexpected error while attempting to execute the program.\n\n" +
-						"The program was executed, however it returned no additional error messaging." +
-						fmt.Sprintf("\n\nProgram: %s", cmd.Path) +
+				resp.Diagnostics.AddAttributeError(
+					path.Root("program"),
+					"External Program Execution Failed",
+					"The data source received an unexpected error while attempting to execute the program."+
+						fmt.Sprintf("\n\nProgram: %s", cmd.Path)+
+						fmt.Sprintf("\nError Message: %s", string(exitErr.Stderr))+
 						fmt.Sprintf("\nState: %s", err),
-					AttributePath: cty.GetAttrPath("program"),
-				},
+				)
+				return
 			}
-		}
 
-		return diag.Diagnostics{
-			{
-				Severity: diag.Error,
-				Summary:  "External Program Execution Failed",
-				Detail: "The data source received an unexpected error while attempting to execute the program." +
-					fmt.Sprintf("\n\nProgram: %s", cmd.Path) +
-					fmt.Sprintf("\nError: %s", err),
-				AttributePath: cty.GetAttrPath("program"),
-			},
+			resp.Diagnostics.AddAttributeError(
+				path.Root("program"),
+				"External Program Execution Failed",
+				"The data source received an unexpected error while attempting to execute the program.\n\n"+
+					"The program was executed, however it returned no additional error messaging."+
+					fmt.Sprintf("\n\nProgram: %s", cmd.Path)+
+					fmt.Sprintf("\nState: %s", err),
+			)
+			return
 		}
+
+		resp.Diagnostics.AddAttributeError(
+			path.Root("program"),
+			"External Program Execution Failed",
+			"The data source received an unexpected error while attempting to execute the program."+
+				fmt.Sprintf("\n\nProgram: %s", cmd.Path)+
+				fmt.Sprintf("\nError: %s", err),
+		)
+		return
 	}
 
 	result := map[string]string{}
 	err = json.Unmarshal(resultJson, &result)
 	if err != nil {
-		return diag.Diagnostics{
-			{
-				Severity: diag.Error,
-				Summary:  "Unexpected External Program Results",
-				Detail: `The data source received unexpected results after executing the program.
+		resp.Diagnostics.AddAttributeError(
+			path.Root("program"),
+			"Unexpected External Program Results",
+			`The data source received unexpected results after executing the program.
 
 Program output must be a JSON encoded map of string keys and string values.
 
 If the error is unclear, the output can be viewed by enabling Terraform's logging at TRACE level. Terraform documentation on logging: https://www.terraform.io/internals/debugging
-` +
-					fmt.Sprintf("\nProgram: %s", cmd.Path) +
-					fmt.Sprintf("\nResult Error: %s", err),
-				AttributePath: cty.GetAttrPath("program"),
-			},
-		}
+`+
+				fmt.Sprintf("\nProgram: %s", cmd.Path)+
+				fmt.Sprintf("\nResult Error: %s", err),
+		)
+		return
 	}
 
-	d.Set("result", result)
+	config.Result, _ = types.MapValueFrom(ctx, types.StringType, result)
```

Comment:
While `map[string]string` should always convert okay given the current code, we should return and handle diagnostics here for defensive programming should the other code change. 👍